src/amt/cim.ts

```diff
@@ -3,56 +3,251 @@
-  ENUMERATE = 'Enumerate',
-  RELEASE = 'Release'
-}
-
-export enum CIM_Actions {
-  ENUMERATE = 'enumeration/Enumerate',
-  PULL = 'enumeration/Pull',
-  GET = 'transfer/Get'
-}
+import { Selector, WSManMessageCreator } from './wsman'
+import { CIM_Methods, CIM_Actions, CIM_Classes } from './enums/cim_enums'
 
 export class CIM {
-  wsmanMessageCreator: WSManMessageCreator
-  resourceUriBase: 'http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/'
-  commonCimRUri: 'http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/'
-  constructor () {
-    this.wsmanMessageCreator = new WSManMessageCreator()
+  wsmanMessageCreator: WSManMessageCreator = new WSManMessageCreator()
+  readonly resourceUriBase: string = 'http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/'
+  private enumerate (action: CIM_Actions, cimClass: CIM_Classes, messageId: String): String {
+    const header: String = this.wsmanMessageCreator.createHeader(action, `${this.resourceUriBase}${cimClass}`, messageId)
+    const body: String = this.wsmanMessageCreator.createBody(CIM_Methods.ENUMERATE)
+    const response: String = this.wsmanMessageCreator.createXml(header, body)
+    return response
+  }
+
+  private pull (action: CIM_Actions, cimClass: CIM_Classes, messageId: String, enumerationContext: String): String {
+    const header: String = this.wsmanMessageCreator.createHeader(action, `${this.resourceUriBase}${cimClass}`, messageId)
+    const body: String = this.wsmanMessageCreator.createBody(CIM_Methods.PULL, enumerationContext)
+    const response: String = this.wsmanMessageCreator.createXml(header, body)
+    return response
+  }
+
+  private get (action: CIM_Actions, cimClass: CIM_Classes, messageId: String): String {
+    const header: String = this.wsmanMessageCreator.createHeader(action, `${this.resourceUriBase}${cimClass}`, messageId)
+    const body: String = this.wsmanMessageCreator.createBody(CIM_Methods.GET)
+    const response: String = this.wsmanMessageCreator.createXml(header, body)
+    return response
   }
 
   cim_ServiceAvailableToElement = (method: CIM_Methods, messageId: String, enumerationContext?: String): String => {
-    let response: String, header: String, body: String
+    switch (method) {
+      case 'Pull':
+        return this.pull(CIM_Actions.PULL, CIM_Classes.CIM_SERVICE_AVAILABLE_TO_ELEMENT, messageId, enumerationContext)
+      case 'Enumerate':
+        return this.enumerate(CIM_Actions.ENUMERATE, CIM_Classes.CIM_SERVICE_AVAILABLE_TO_ELEMENT, messageId)
+      default:
+        return null
+    }
+  }
+
+  cim_SoftwareIdentity = (method: CIM_Methods, messageId: String, enumerationContext?: String): String => {
+    switch (method) {
+      case 'Enumerate':
+        return this.enumerate(CIM_Actions.ENUMERATE, CIM_Classes.CIM_SOFTWARE_IDENTITY, messageId)
+      case 'Pull':
+        if (enumerationContext != null) {
+          return this.pull(CIM_Actions.PULL, CIM_Classes.CIM_SOFTWARE_IDENTITY, messageId, enumerationContext)
+        } else {
```

Comment:
else is unnecessary, could even do a fallthrough statement to execute the default: but some people don't prefer that.

Reply:
updated